### ISSUE TITLE
semgrep timeout error "python.lang.maintainability.useless-literal.useless-literal-dict" on big dictionnaries

### ISSUE THREAD

**Issue by chrisdlangton** [semgrep/semgrep#2663]
**Describe the bug**
Large dict causes it 100% reproducable

**Expected behavior**
Large dict do not fail/terminate the execution

I am using `--timeout-threshold=0` and the docs say `If set to 0 will not have limit`

**Comment by ievans** [comment]
thanks for the bug report @chrisdlangton!

What version of semgrep are you on? `semgrep --version`? I think we fixed that timeout not being honored recently.

do you by chance have a reproducing test case? You can share here or in the playground https://semgrep.dev/editor

finally, what's the priority of the issue to you? is it blocking you from using Semgrep further, or just an annoyance?

**Comment by chrisdlangton** [comment]
latest at time of posting `0.41.1` doing a `pip install -U semgrep` the version remained
didn't have a test, can't really think of a fictitious dictionary of at least 50k keys... can you? how would I even generate one? I have mine created by hand and would not wish that pain on anyone... any tips on generalisation of a large python dict? Is web based even going to support that?
Priority is minimal, we were evaluating semgrep for our micro severvises pipelines and currently use a bunch of linters and code quality tools (pylint, black, sonarlint, etc.) We were looking to consolidate and hopefully get rid of sonarlint entirely (but still unsure about the semgrep rules to use as we found no mapping to sonarlint's repository of rules)

**Comment by chrisdlangton** [comment]
i created this 
https://semgrep.dev/s/chrisdlangton:py-useless-literal
seems to be borked (as expected)
the rule is straight out of 
https://semgrep.dev/p/r2c-ci
and the key `2` in the dict should match that rule and semgrep seems to bork on large dicts generally (but this is a good simple rule to cause the issue.

**NOTE**
by the way, the playground is totally artificial for this bug, it is irrelevant
why? because it is not allowing a demonstration of the  `--timeout-threshold` argument, period.

**Comment by ievans** [comment]
Thanks so much for the repro, that's super helpful. @aryx or @mjambon maybe add this case to the profiling suite work and we can track it before fix -> after? also we probably need a separate issue about `--timeout-threshold` not being honored?

@chrisdlangton we don't have a 1:1 mapping with sonarlint rules, though we have done that for some other OSS scanners. Making it easier to discover/fork/remix rules is high on the priority list (thanks for sharing your relative priority). cc @DrewDennison. Currently it is more optimized for writing your own rules.

Appreciate your patience and if you have more feedback on the UX from your evaluation, we're all ears here or via [email](mailto:support@r2c.dev) or [slack](https://r2c.dev/slack)!

**Comment by colleend** [comment]
Hi! @chrisdlangton did you happen to want the functionality ```--timeout``` instead of ```--timeout-threshold```? 

```--timeout``` indicates the maximum time spent running the rule on a single file, and ```--timeout-threshold``` indicates the max number of rules that can timeout on a file. 

Using ```--timeout-threshold 0``` resulted in the predicted timeout error on the file, and using ```--timeout 0``` resulted in the following error:
```
running 1 rules...
semgrep-core exit code: -11
unexpected non-json output while invoking semgrep-core:
--- semgrep-core stdout ---

--- end semgrep-core stdout ---
--- semgrep-core stderr ---

--- end semgrep-core stderr ---
An error occurred while invoking the semgrep engine; please help us fix this by creating an issue at https://github.com/returntocorp/semgrep
```

I'll look into the semgrep-core invoke-error a bit more tomorrow -- it may be a PA fix, but I'll take a glance for sure.

**Comment by chrisdlangton** [comment]
@colleend that tip does work as described.

I have to be completely honest, I don't want any timeout to cause failure. If I have a large dictionary the tests should still run and pass if there are no rule violations regardless of size. 

using timeout settings are brittle, and require tuning as the code base changes and the semgrep config increases too - the most important thing is the timeout solves for a symptom and not a root cause.
I'd like there to be a solution that resolves the root cause of the failures linked to large dictionaries

**Comment by colleend** [comment]
Yes definitely -- thanks for your fast reply! What would you say is your priority for this issue? It might be a tough fix.

I've tried digging a little deeper into it, and it seems like the command
 
``` semgrep-core/_build/default/cli/Main.exe -lang python -e '{..., $X: $A, ..., $X: $B, ...}' dict-timeout-test.py```

where ```dict-timeout-test.py``` is the dictionary provided in semgrep-dev link results in an error:

```Your system has run out of application memory```

This is most likely because the matching process (especially with the metavariables and ellipsis both included) requires storage of everything that $X could be, and the dictionary is so large that this takes up too much memory. I'm going to tag @aryx and @mjambon about this issue, as they are our semgrep-core PA experts 😄

**Comment by dlukeomalley** [comment]
@colleend Cleaning up and organizing tickets for upcoming sprints. My read of the above is that this is a performance issue and there may be other tickets to file - are there any other tickets you can link to for us to prioritize? Thank you! 🙏

**Comment by colleend** [comment]
Currently no other tickets linked to this issue -- I would say that this is a lower priority performance issue due to the potential complexity of the fix & the rarity of scanning such large dicts.

**Comment by IagoAbal** [comment]
This alone already times out https://semgrep.dev/s/d2K0.

**Comment by IagoAbal** [comment]
AC matching for example tries to detect these situations heuristically, and if so it enters a "degraded mode". We could perhaps do something similar here, although one could argue that Semgrep is not meant to analyze this kind of machine-generated code... in fact we have plans to skip this kind of code automatically.

**Comment by IagoAbal** [comment]
But I would expect this https://semgrep.dev/s/Z4v4 to work and be relatively fast, perhaps we could try to optimize range intersection (i.e., `Range_with_metavars.intersect_ranges`).

**Comment by ievans** [comment]
Still an issue
